src/overlays/actors/ovl_En_Warp_tag/z_en_warp_tag.c

```diff
@@ -34,32 +36,249 @@ const ActorInit En_Warp_tag_InitVars = {
+
+/**
+ * Loads DynaPoly from GAMEPLAY_DANGEON_KEEP.
+ */
+void EnWarpTag_CheckDungeonKeepObject(EnWarptag* this, GlobalContext* globalCtx) {
+    if (Object_IsLoaded(&globalCtx->objectCtx, this->dangeonKeepObject)) {
+        this->actionFunc = EnWarpTag_WaitForPlayer;
+        DynaPolyActor_Init(&this->dyna, 0x1);
+        DynaPolyActor_LoadMesh(globalCtx, &this->dyna, &gWarpTagGoronTrialBaseCollider);
+        this->dyna.actor.objBankIndex = this->dangeonKeepObject;
+        this->dyna.actor.draw = EnWarpTag_Draw;
+    }
+}
+
+void EnWarpTag_WaitForPlayer(EnWarptag* this, GlobalContext* globalCtx) {
+    if (!Player_InCsMode(&globalCtx->state) && (this->dyna.actor.xzDistToPlayer <= 30.0f) &&
+        (this->dyna.actor.playerHeightRel <= 10.0f)) {
+        if (GET_WARPTAG_INVISIBLE(this)) {
+            func_800B7298(globalCtx, NULL, 0x51);
+            this->actionFunc = EnWarpTag_GrottoReturn;
+        } else {
+            func_800B7298(globalCtx, NULL, 0xF);
+            this->actionFunc = EnWarpTag_RespawnPlayer;
+        }
+    }
+}
+
+/**
+ * Unused ActionFunc: assigned in EnWarpTag_Init, no known variants use.
+ */
+void EnWarpTag_Unused809C09A0(EnWarptag* this, GlobalContext* globalCtx) {
+    if (func_800B8718(&this->dyna.actor, &globalCtx->state)) {
+        // func above: checks for ACTOR_FLAG_20000000, returns true and resets if set, else return false
+        //   this actor doesnt have that flag set default, or in init, and this is called shortly after init
+        //   and I doubt its set externally by another actor, so I believe this is unused
+        // might be a bug, they might have meant to set actor flag (0x2000 0000) up above but mistyped (0x200 0000)
+        // also GET_WARPTAG_3C0 should always return 2C0 -> 0xF for all known in-game uses, which is OOB
+        func_80152434(globalCtx, D_809C1000[GET_WARPTAG_3C0(this)]); // unk message function
+        this->actionFunc = EnWarpTag_Unused809C0A20;
+
+    } else {
+        func_800B8804(&this->dyna.actor, globalCtx, 50.0f); // updates player->unk_A90
+    }
+}
 
-extern InitChainEntry D_809C1008[];
+/**
+ * Unused ActionFunc: assigned by EnWarpTag_Unused809C09A0, no known variants use.
+ */
+void EnWarpTag_Unused809C0A20(EnWarptag* this, GlobalContext* globalCtx) {
+    if (globalCtx->msgCtx.ocarinaMode == 9) {
+        func_800B7298(globalCtx, NULL, 7);
+        this->actionFunc = EnWarpTag_RespawnPlayer;
+        ActorCutscene_Stop(ActorCutscene_GetCurrentIndex());
+
+    } else if (globalCtx->msgCtx.ocarinaMode >= 2) {
+        globalCtx->msgCtx.ocarinaMode = 4;
+        this->actionFunc = EnWarpTag_Unused809C09A0;
+    }
+}
+
+/**
+ * ActionFunc: Goron Trial (Moon), respawn at the beginning of goron rolling track, try again.
+ */
+void EnWarpTag_RespawnPlayer(EnWarptag* this, GlobalContext* globalCtx) {
+    ActorEntry* playerActorEntry;
+    Player* player;
+    s32 playerSpawnIndex;
+    s32 new15E;
+    s32 entranceIndex;
+    u32 playerSpawnIndexPerForm[5];
```

Comment:
I think you could use `PLAYER_FORM_MAX`  here
```suggestion
    u32 playerSpawnIndexPerForm[PLAYER_FORM_MAX];
```